Includes externally applied forces forces in MBT::CalcInverseDynamics().

amcastro-tri opened the pull request:
Essentially the API (and implementation) for `CalcInverseDynamics()` is updated to accept input vectors of applied forces.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/robotlocomotion/drake/6958)
<!-- Reviewable:end -->

amcastro-tri commented:
Reviewed 14 of 16 files at r1.
Review status: 14 of 16 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6958)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
Reviewed 2 of 5 files at r2.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6958)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
+@siyuanfeng-tri for feature review please.

---

Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6958#-:-KszrCESC1iYnzKBg2sT:bakb6i6)*
<!-- Sent from Reviewable.io -->

siyuanfeng-tri commented:
first pass done, no major complaints. nice. 

---

Reviewed 8 of 16 files at r1, 2 of 5 files at r2, 2 of 3 files at r3.
Review status: 12 of 13 files reviewed at latest revision, 6 unresolved discussions.

---

*[drake/multibody/benchmarks/acrobot/acrobot.cc, line 143 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/6958#-Kt7X_7ILd46J_aDYtVR:-Kt7X_7ILd46J_aDYtVS:b2moxsj) ([raw file](https://github.com/robotlocomotion/drake/blob/adb7c87ffcec99106151f38807472861b0a9f9e8/drake/multibody/benchmarks/acrobot/acrobot.cc#L143)):*
> ```cpp
>   const Isometry3<T> X_WLcm = CalcLink2PoseInWorldFrame(theta1, theta2);
>   // Pose of the elbow outboard frame Eo in Lcm's frame.
>   const Isometry3<T> X_LcmEo(Translation3<T>(0.0, lc2_, 0.0));
> ```

BTW is it -lc2_ or +? idk how the frames are defined here. 

---

*[drake/multibody/benchmarks/acrobot/acrobot.h, line 76 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/6958#-Kt7XIqdopFMjC2y-LOh:-Kt7XIqdopFMjC2y-LOi:bqwn0cp) ([raw file](https://github.com/robotlocomotion/drake/blob/adb7c87ffcec99106151f38807472861b0a9f9e8/drake/multibody/benchmarks/acrobot/acrobot.h#L76)):*
> ```objc
> 
>   /// Computes the the term `G(q)` containing the effect of gravity as a
>   /// function of the state of the pendulum.
> ```

BTW state -> configuration?

---

*[drake/multibody/multibody_tree/body_node.h, line 578 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/6958#-Kt7jmwStGae2HOj6Hpx:-Kt7jmwStGae2HOj6Hpy:b-neadff) ([raw file](https://github.com/robotlocomotion/drake/blob/adb7c87ffcec99106151f38807472861b0a9f9e8/drake/multibody/multibody_tree/body_node.h#L578)):*
> ```objc
>   ///   Externally applied generalized force at this node's mobilizer. It can
>   ///   have zero size, implying no generalized forces are applied. Otherwise it
>   ///   must have a size equal to the number of generalized coordinates for this
> ```

BTW coordinate -> velocity? 

---

*[drake/multibody/multibody_tree/mobilizer.h, line 431 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/6958#-Kt7c6B_Q_-GM0207eQ9:-Kt7c6B_Q_-GM0207eQA:bad4ih4) ([raw file](https://github.com/robotlocomotion/drake/blob/adb7c87ffcec99106151f38807472861b0a9f9e8/drake/multibody/multibody_tree/mobilizer.h#L431)):*
> ```objc
>   /// for `this` mobilizer from a vector of generalized velocities for the
>   /// entire MultibodyTree model.
>   /// @note This same method can be used to access arrays of generalized
> ```

BTW i would just explicitly add the functions to avoid confusion. 

---

*[drake/multibody/multibody_tree/multibody_tree.h, line 652 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/6958#-Kt7dVF7YgktuMxxRhJ8:-Kt7dVF7YgktuMxxRhJ9:b5biaet) ([raw file](https://github.com/robotlocomotion/drake/blob/adb7c87ffcec99106151f38807472861b0a9f9e8/drake/multibody/multibody_tree/multibody_tree.h#L652)):*
> ```objc
>   ///   the mutable version of this method to write into this array.
>   /// @param[in] Fapplied_Bo_W_array
>   ///   A pointer to a vector containing the spatial force `Fapplied_Bo_W`
> ```

pointer -> ref.
BTW, can you do like a unordered_map<index, spatial_force>? 

---

*[drake/multibody/multibody_tree/math/spatial_force.h, line 147 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/6958#-Kt7ZKf0KnhnFtWUJwGu:-Kt7ZKf0KnhnFtWUJwGv:b-mos2vh) ([raw file](https://github.com/robotlocomotion/drake/blob/adb7c87ffcec99106151f38807472861b0a9f9e8/drake/multibody/multibody_tree/math/spatial_force.h#L147)):*
> ```objc
>   }
> 
>   /// Subtracts a spatial force from `this` spatial force.
> ```

BTW, it would be extra nice to have some mechanism check this, the assumptions are really easily violated.. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6958#-:-Kt7n40xwJ21aeDxbver:b-u33ktd)*
<!-- Sent from Reviewable.io -->

siyuanfeng-tri commented:
Reviewed 1 of 3 files at r3.
Review status: all files reviewed at latest revision, 10 unresolved discussions.

---

*[drake/multibody/multibody_tree/test/double_pendulum_test.cc, line 595 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/6958#-Kt8gtTKASZYqCkiRAOl:-Kt8gtTKASZYqCkiRAOm:b-shb53x) ([raw file](https://github.com/robotlocomotion/drake/blob/adb7c87ffcec99106151f38807472861b0a9f9e8/drake/multibody/multibody_tree/test/double_pendulum_test.cc#L595)):*
> ```cpp
>     // the data given at input is lost on output. A user might choose then to
>     // have separate input/output arrays.
>     VectorXd vdot(model_->get_num_velocities());
> ```

probably need to set this to 0?

---

*[drake/multibody/multibody_tree/test/double_pendulum_test.cc, line 875 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/6958#-Kt8hUNEE35dHw-fag2d:-Kt8hUNEE35dHw-fag2e:b-2sk11e) ([raw file](https://github.com/robotlocomotion/drake/blob/adb7c87ffcec99106151f38807472861b0a9f9e8/drake/multibody/multibody_tree/test/double_pendulum_test.cc#L875)):*
> ```cpp
> 
>       EXPECT_TRUE(V_WU.IsApprox(V_WU_expected, kTolerance));
>       EXPECT_TRUE(V_WLcm.IsApprox(V_WL_expected, kTolerance));
> ```

BTW might want to change V_WL_expected to V_WLcm, the benchmark's L is at Lcm? 

---

*[drake/multibody/multibody_tree/test/double_pendulum_test.cc, line 907 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/6958#-Kt8hiPV0hxxtH3Ua4LO:-Kt8hiPV0hxxtH3Ua4LP:brx2yk6) ([raw file](https://github.com/robotlocomotion/drake/blob/adb7c87ffcec99106151f38807472861b0a9f9e8/drake/multibody/multibody_tree/test/double_pendulum_test.cc#L907)):*
> ```cpp
> 
>       EXPECT_TRUE(A_WU.IsApprox(A_WU_expected, kTolerance));
>       EXPECT_TRUE(A_WLcm.IsApprox(A_WL_expected, kTolerance));
> ```

same as above. 

---

*[drake/multibody/multibody_tree/test/double_pendulum_test.cc, line 935 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/6958#-Kt8hnLcdf0G5Mbeblaq:-Kt8hnLcdf0G5Mbeblar:bb7od0r) ([raw file](https://github.com/robotlocomotion/drake/blob/adb7c87ffcec99106151f38807472861b0a9f9e8/drake/multibody/multibody_tree/test/double_pendulum_test.cc#L935)):*
> ```cpp
> 
>       EXPECT_TRUE(A_WU.IsApprox(A_WU_expected, kTolerance));
>       EXPECT_TRUE(A_WLcm.IsApprox(A_WL_expected, kTolerance));
> ```

same

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6958)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
Review status: all files reviewed at latest revision, 6 unresolved discussions.

---

*[drake/multibody/multibody_tree/multibody_tree.h, line 652 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/6958#-Kt7dVF7YgktuMxxRhJ8:-KtHN-5qTH9RUvqbazJm:b-h20ozv) ([raw file](https://github.com/robotlocomotion/drake/blob/adb7c87ffcec99106151f38807472861b0a9f9e8/drake/multibody/multibody_tree/multibody_tree.h#L652)):*
Good catch. 
> can you do like a unordered_map<index, spatial_force>?

I considered this option. The reason I am using an array here is because we'll have an additional layer on top of `MultibodyTree` that will provide this additional sugar.  That gives us freedom to keep `MultibodyTree` primitive, simple, fast.

---

*[drake/multibody/multibody_tree/math/spatial_force.h, line 147 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/6958#-Kt7ZKf0KnhnFtWUJwGu:-KtHNRuuapzAodVJYwGC:b1nb9cn) ([raw file](https://github.com/robotlocomotion/drake/blob/adb7c87ffcec99106151f38807472861b0a9f9e8/drake/multibody/multibody_tree/math/spatial_force.h#L147)):*
<details><summary><i>Previously, siyuanfeng-tri wrote…</i></summary><blockquote>

BTW, it would be extra nice to have some mechanism check this, the assumptions are really easily violated.. 
</blockquote></details>

BTW you mean the frames? yes, that'd be super nice! however all this spatial algebra classes only provide the math, not frame checking or anything of the like. Just fast, verified algebra to build MBT on top of it. I know Hongkai and Paul are super interested in implementing something like this. Unfortunately time is finite and its not my top priority right now.
So far, even when simple, they proved super useful (vs the alternative of using Eigen vectors and matrices) since there is type checking, they are super nice as containers of useful methods for frame shifting, re-expression, etc and they offer a lot more of semantic sugar already. Also the templatization gets easier, vs the alternative of having heavily templated free methods taking those awful `MatrixBase<Derived>`

---

*[drake/multibody/multibody_tree/test/double_pendulum_test.cc, line 595 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/6958#-Kt8gtTKASZYqCkiRAOl:-KtHP3fTwp0olIW0bC-B:buf54yl) ([raw file](https://github.com/robotlocomotion/drake/blob/adb7c87ffcec99106151f38807472861b0a9f9e8/drake/multibody/multibody_tree/test/double_pendulum_test.cc#L595)):*
<details><summary><i>Previously, siyuanfeng-tri wrote…</i></summary><blockquote>

probably need to set this to 0?
</blockquote></details>

good catch. Done.

---

*[drake/multibody/multibody_tree/test/double_pendulum_test.cc, line 875 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/6958#-Kt8hUNEE35dHw-fag2d:-KtHQGRaqq37IaZ6m2nW:b-rhhf6a) ([raw file](https://github.com/robotlocomotion/drake/blob/adb7c87ffcec99106151f38807472861b0a9f9e8/drake/multibody/multibody_tree/test/double_pendulum_test.cc#L875)):*
<details><summary><i>Previously, siyuanfeng-tri wrote…</i></summary><blockquote>

BTW might want to change V_WL_expected to V_WLcm, the benchmark's L is at Lcm? 
</blockquote></details>

BTW fair enough. Done.

---

*[drake/multibody/multibody_tree/test/double_pendulum_test.cc, line 907 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/6958#-Kt8hiPV0hxxtH3Ua4LO:-KtHQJqEW0ayE9pPlPUb:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/adb7c87ffcec99106151f38807472861b0a9f9e8/drake/multibody/multibody_tree/test/double_pendulum_test.cc#L907)):*
<details><summary><i>Previously, siyuanfeng-tri wrote…</i></summary><blockquote>

same as above. 
</blockquote></details>

Done.

---

*[drake/multibody/multibody_tree/test/double_pendulum_test.cc, line 935 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/6958#-Kt8hnLcdf0G5Mbeblaq:-KtHQLEyeI8TqEzgCPof:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/adb7c87ffcec99106151f38807472861b0a9f9e8/drake/multibody/multibody_tree/test/double_pendulum_test.cc#L935)):*
<details><summary><i>Previously, siyuanfeng-tri wrote…</i></summary><blockquote>

same
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6958)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
+@soonho-tri for platform review please.

---

Review status: 6 of 13 files reviewed at latest revision, 4 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6958#-:-KtHQa3B20ObSiPIY1VH:b-vt9vbo)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
+@sherm1 for a quick feature review pass.

---

Reviewed 3 of 16 files at r1, 1 of 5 files at r2, 6 of 7 files at r4.
Review status: 12 of 13 files reviewed at latest revision, 8 unresolved discussions.

---

*[drake/common/eigen_types.h, line 51 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/6958#-KtH_QQ_KdfTC2SFi-XB:-KtH_QQ_KdfTC2SFi-XC:b-f6e00e) ([raw file](https://github.com/robotlocomotion/drake/blob/78544728340f8de2c30e5c52edd73f9ec81b905c/drake/common/eigen_types.h#L51)):*
> ```objc
> /// elements.
> template <typename Scalar>
> using VectorUpTo6 = Eigen::Matrix<Scalar, Eigen::Dynamic, 1, 0, 6, 1>;
> ```

BTW great idea -- I believe these are as fast as fixed-size.

---

*[drake/multibody/benchmarks/acrobot/acrobot.h, line 76 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/6958#-KtHX-1hV_ydYaNtMb4p:-KtHX-1hV_ydYaNtMb4q:bsww87s) ([raw file](https://github.com/robotlocomotion/drake/blob/78544728340f8de2c30e5c52edd73f9ec81b905c/drake/multibody/benchmarks/acrobot/acrobot.h#L76)):*
> ```objc
> 
>   /// Computes the the term `G(q)` containing the effect of gravity as a
>   /// function of the configuration of the pendulum.
> ```

BTW can you be more precise about this? "effect of gravity" is ambiguous. I *think* this is the effective joint-space torques induced by gravity, that is f=JᵀG(q) where G(q) is the gravitational body forces?

---

*[drake/multibody/multibody_tree/multibody_tree.cc, line 319 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/6958#-KtHae19H8levdj_A0Om:-KtHae19H8levdj_A0On:bazogyb) ([raw file](https://github.com/robotlocomotion/drake/blob/78544728340f8de2c30e5c52edd73f9ec81b905c/drake/multibody/multibody_tree/multibody_tree.cc#L319)):*
> ```cpp
>       }
>       if (Fapplied_size != 0) {
>         Fapplied_Bo_W = (*F_BMo_W_array)[body_node_index];
> ```

Why is F_BMo being copied into F_BBo? I would have expected this to be copying from Fapplied_Bo. (If that's right, is there a missing unit test that should fail?)

---

*[drake/multibody/multibody_tree/multibody_tree.cc, line 327 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/6958#-KtHZyvQlAot0pZpRAGf:-KtHZyvQlAot0pZpRAGg:bvettys) ([raw file](https://github.com/robotlocomotion/drake/blob/78544728340f8de2c30e5c52edd73f9ec81b905c/drake/multibody/multibody_tree/multibody_tree.cc#L327)):*
> ```cpp
>           mbt_context, pc, vc, *A_WB_array,
>           Fapplied_Bo_W, tau_applied_mobilizer,
>           F_BMo_W_array, tau_array);
> ```

BTW this notation is really working well! Very clear.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6958#-:-KtHbOtKNdnrpW3iDV1s:b-sbuvzs)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
Reviewed 2 of 16 files at r1, 1 of 7 files at r4.
Review status: all files reviewed at latest revision, 12 unresolved discussions.

---

*[drake/multibody/multibody_tree/multibody_tree.h, line 734 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/6958#-KtHeGnhcuwjnrrwNCGt:-KtHeGnhcuwjnrrwNCGu:bsrnk7o) ([raw file](https://github.com/robotlocomotion/drake/blob/78544728340f8de2c30e5c52edd73f9ec81b905c/drake/multibody/multibody_tree/multibody_tree.h#L734)):*
> ```objc
>       std::vector<SpatialAcceleration<T>>* A_WB_array,
>       std::vector<SpatialForce<T>>* F_BMo_W_array,
>       Eigen::Ref<VectorX<T>> tau_array) const;
> ```

This should be an [EigenPtr](https://github.com/RobotLocomotion/drake/blob/master/drake/common/eigen_types.h#L266) so that the API takes `&tau` rather than just `tau`.

---

*[drake/multibody/multibody_tree/test/double_pendulum_test.cc, line 599 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/6958#-KtHctcTNz0n6sIW1m64:-KtHctcTNz0n6sIW1m65:b-hnokha) ([raw file](https://github.com/robotlocomotion/drake/blob/78544728340f8de2c30e5c52edd73f9ec81b905c/drake/multibody/multibody_tree/test/double_pendulum_test.cc#L599)):*
> ```cpp
>     model_->CalcInverseDynamics(
>         *context_, pc, vc, vdot, F_Bo_W_array, tau,
>         &A_WB_array, &F_Bo_W_array, tau);
> ```

This should be `&tau`. See class [EigenPtr](https://github.com/RobotLocomotion/drake/blob/master/drake/common/eigen_types.h#L266) in eigen_types.h to make this work.

---

*[drake/multibody/multibody_tree/test/double_pendulum_test.cc, line 691 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/6958#-KtHek3KOYUTQuik0ysc:-KtHek3KOYUTQuik0ysd:bfu2cmx) ([raw file](https://github.com/robotlocomotion/drake/blob/78544728340f8de2c30e5c52edd73f9ec81b905c/drake/multibody/multibody_tree/test/double_pendulum_test.cc#L691)):*
> ```cpp
>     vector<SpatialForce<double>> F_BMo_W_array(model_->get_num_bodies());
>     model_->CalcInverseDynamics(*context_, pc, vc, vdot, {}, VectorXd(),
>                                 &A_WB_array, &F_BMo_W_array, tau);
> ```

API should require `&tau` here.

---

*[drake/multibody/multibody_tree/test/double_pendulum_test.cc, line 866 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/6958#-KtHf6ELi9F9IjYQyZD_:-KtHf6ELi9F9IjYQyZDa:bi5u5mb) ([raw file](https://github.com/robotlocomotion/drake/blob/78544728340f8de2c30e5c52edd73f9ec81b905c/drake/multibody/multibody_tree/test/double_pendulum_test.cc#L866)):*
> ```cpp
>       const SpatialVelocity<double> V_WLcm = V_WL.Shift(p_LoLcm_W);
> 
>       const SpatialVelocity<double> V_WUcmcm_expected(
> ```

BTW why "cmcm" here?

---

*[drake/multibody/multibody_tree/test/double_pendulum_test.cc, line 1122 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/6958#-KtHfTiZ_ErNCGtXlx9X:-KtHfTiZ_ErNCGtXlx9Y:bmoh84r) ([raw file](https://github.com/robotlocomotion/drake/blob/78544728340f8de2c30e5c52edd73f9ec81b905c/drake/multibody/multibody_tree/test/double_pendulum_test.cc#L1122)):*
> ```cpp
>           const SpatialVelocity<double> V_WLcm = V_WL.Shift(p_LoLcm_W);
> 
>           const SpatialVelocity<double> V_WUcmcm_expected(
> ```

BTW cmcm?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6958)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
Review status: all files reviewed at latest revision, 13 unresolved discussions.

---

*[drake/multibody/multibody_tree/multibody_tree.h, line 714 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/6958#-KtHjoBKAmKCu5baMLfX:-KtHjoBKAmKCu5baMLfY:b-oz4f3n) ([raw file](https://github.com/robotlocomotion/drake/blob/78544728340f8de2c30e5c52edd73f9ec81b905c/drake/multibody/multibody_tree/multibody_tree.h#L714)):*
> ```objc
>   /// @note `F_BMo_W_array` (`tau_array`) and `Fapplied_Bo_W_array`
>   /// (`tau_applied_array`) can actually be the same
>   /// array in order to reduce memory footprint and/or dynamic memory
> ```

BTW I like that the same array can be used, mostly because users might do it anyway and it is better that it works than that we have to explain to them why it doesn't! However there should be a unit test that validates that the code works the same way with disjoint memory as with overwritten memory. I didn't see one.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6958)*
<!-- Sent from Reviewable.io -->

siyuanfeng-tri commented:
Reviewed 6 of 7 files at r4.
Review status: all files reviewed at latest revision, 12 unresolved discussions.

---

*[drake/multibody/multibody_tree/mobilizer.h, line 453 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/6958#-KtHmS_RN1DzpTLfLaZ6:-KtHmS_S_jfnFNxpVKcm:bpk19ih) ([raw file](https://github.com/robotlocomotion/drake/blob/78544728340f8de2c30e5c52edd73f9ec81b905c/drake/multibody/multibody_tree/mobilizer.h#L453)):*
> ```objc
>   /// MultibodyTree::get_num_velocities().
>   Eigen::VectorBlock<const Eigen::Ref<const VectorX<T>>>
>   get_forces_from_array(const Eigen::Ref<const VectorX<T>>& tau_array) const {
> ```

BTW why get rid of genalized_? 

---

*[drake/multibody/multibody_tree/multibody_tree.cc, line 319 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/6958#-KtHae19H8levdj_A0Om:-KtHnLwBvM2Eq2xiPsX1:byon0yq) ([raw file](https://github.com/robotlocomotion/drake/blob/78544728340f8de2c30e5c52edd73f9ec81b905c/drake/multibody/multibody_tree/multibody_tree.cc#L319)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

Why is F_BMo being copied into F_BBo? I would have expected this to be copying from Fapplied_Bo. (If that's right, is there a missing unit test that should fail?)
</blockquote></details>

Fapplied_Bo_W_array??

---

*[drake/multibody/multibody_tree/multibody_tree.h, line 734 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/6958#-KtHeGnhcuwjnrrwNCGt:-KtHncVmoCYsA7BhC1H_:b-2o5113) ([raw file](https://github.com/robotlocomotion/drake/blob/78544728340f8de2c30e5c52edd73f9ec81b905c/drake/multibody/multibody_tree/multibody_tree.h#L734)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

This should be an [EigenPtr](https://github.com/RobotLocomotion/drake/blob/master/drake/common/eigen_types.h#L266) so that the API takes `&tau` rather than just `tau`.
</blockquote></details>

hmmm, i think we have always being using Eigen::Ref for mutable stuff, (at least i know i have been. )

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6958)*
<!-- Sent from Reviewable.io -->

siyuanfeng-tri commented:
Reviewed 1 of 7 files at r4.
Review status: all files reviewed at latest revision, 12 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6958)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
Review status: all files reviewed at latest revision, 12 unresolved discussions.

---

*[drake/multibody/multibody_tree/multibody_tree.h, line 734 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/6958#-KtHeGnhcuwjnrrwNCGt:-KtHot8JWH1RcUqrsNrI:b-80ql1i) ([raw file](https://github.com/robotlocomotion/drake/blob/78544728340f8de2c30e5c52edd73f9ec81b905c/drake/multibody/multibody_tree/multibody_tree.h#L734)):*
<details><summary><i>Previously, siyuanfeng-tri wrote…</i></summary><blockquote>

hmmm, i think we have always being using Eigen::Ref for mutable stuff, (at least i know i have been. )
</blockquote></details>

Yes, this is new @siyuanfeng-tri . See issue #6804 and related PR #6852.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6958)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---

*[drake/multibody/multibody_tree/mobilizer.h, line 453 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/6958#-KtHmS_RN1DzpTLfLaZ6:-KtHqmb_6qAnedWK_tP7:brglpcd) ([raw file](https://github.com/robotlocomotion/drake/blob/78544728340f8de2c30e5c52edd73f9ec81b905c/drake/multibody/multibody_tree/mobilizer.h#L453)):*
<details><summary><i>Previously, siyuanfeng-tri wrote…</i></summary><blockquote>

BTW why get rid of genalized_? 
</blockquote></details>

BTW it was making these methods names way too long IMO. What do you think? especially for the mutable versions like `get_mutable_generalized_forces_from_array()`...

---

*[drake/multibody/multibody_tree/multibody_tree.cc, line 319 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/6958#-KtHae19H8levdj_A0Om:-KtHgosB-ayQyXsCGZCR:b-ymg49o) ([raw file](https://github.com/robotlocomotion/drake/blob/78544728340f8de2c30e5c52edd73f9ec81b905c/drake/multibody/multibody_tree/multibody_tree.cc#L319)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

Why is F_BMo being copied into F_BBo? I would have expected this to be copying from Fapplied_Bo. (If that's right, is there a missing unit test that should fail?)
</blockquote></details>

Nice catch @sherm1! good eye!. I was trying to be overly smart in my previous version to reduce memory footprint and this is a leftover. Fixed and added unit test (which did not pass without the fix).

---

*[drake/multibody/multibody_tree/multibody_tree.cc, line 327 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/6958#-KtHZyvQlAot0pZpRAGf:-KtHh0oqDrJAN4_TpLHj:bd7ukws) ([raw file](https://github.com/robotlocomotion/drake/blob/78544728340f8de2c30e5c52edd73f9ec81b905c/drake/multibody/multibody_tree/multibody_tree.cc#L327)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW this notation is really working well! Very clear.
</blockquote></details>

BTW. Your idea.

---

*[drake/multibody/multibody_tree/multibody_tree.h, line 714 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/6958#-KtHjoBKAmKCu5baMLfX:-KtHr17js2qTdojjuaN2:bae8qpl) ([raw file](https://github.com/robotlocomotion/drake/blob/78544728340f8de2c30e5c52edd73f9ec81b905c/drake/multibody/multibody_tree/multibody_tree.h#L714)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW I like that the same array can be used, mostly because users might do it anyway and it is better that it works than that we have to explain to them why it doesn't! However there should be a unit test that validates that the code works the same way with disjoint memory as with overwritten memory. I didn't see one.
</blockquote></details>

BTW Done per bug fix!

---

*[drake/multibody/multibody_tree/multibody_tree.h, line 734 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/6958#-KtHeGnhcuwjnrrwNCGt:-KtHq1ZWk6_AEOHTLB54:b-7fblx1) ([raw file](https://github.com/robotlocomotion/drake/blob/78544728340f8de2c30e5c52edd73f9ec81b905c/drake/multibody/multibody_tree/multibody_tree.h#L734)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

Yes, this is new @siyuanfeng-tri . See issue #6804 and related PR #6852.
</blockquote></details>

Done. See DRAKE_DEMAND in the cc file. It'd be nice to be able to say `DRAKE_DEMAND(tau_array != nullptr)` @ericcousineau-tri 

---

*[drake/multibody/multibody_tree/test/double_pendulum_test.cc, line 599 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/6958#-KtHctcTNz0n6sIW1m64:-KtHqJo5GsEjkS7XS-K0:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/78544728340f8de2c30e5c52edd73f9ec81b905c/drake/multibody/multibody_tree/test/double_pendulum_test.cc#L599)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

This should be `&tau`. See class [EigenPtr](https://github.com/RobotLocomotion/drake/blob/master/drake/common/eigen_types.h#L266) in eigen_types.h to make this work.
</blockquote></details>

Done.

---

*[drake/multibody/multibody_tree/test/double_pendulum_test.cc, line 691 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/6958#-KtHek3KOYUTQuik0ysc:-KtHqL48j9j54__Xo9mq:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/78544728340f8de2c30e5c52edd73f9ec81b905c/drake/multibody/multibody_tree/test/double_pendulum_test.cc#L691)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

API should require `&tau` here.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6958)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/> with some minor stuff to discuss.

---

Reviewed 7 of 7 files at r5.
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[drake/common/eigen_types.h, line 238 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/6958#-KtI-itWpePA-wKj26zK:-KtI-itWpePA-wKj26zL:bgs24p0) ([raw file](https://github.com/robotlocomotion/drake/blob/75c47ffe55f109bd1d7d68a9e0878843c14dd37a/drake/common/eigen_types.h#L238)):*
> ```objc
> 
> /// This wrapper class provides a way to write non-template functions taking raw
> /// pointers to Eigen objects as parmameters while limiting the number of copies,
> ```

BTW typo

---

*[drake/multibody/multibody_tree/multibody_tree.h, line 624 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/6958#-KtHwEA46MVr3HUk-sqr:-KtHwEA46MVr3HUk-sqs:blyhqd0) ([raw file](https://github.com/robotlocomotion/drake/blob/75c47ffe55f109bd1d7d68a9e0878843c14dd37a/drake/multibody/multibody_tree/multibody_tree.h#L624)):*
> ```objc
>   /// Mobilizer in order to attain the specified generalized accelerations.
>   /// Mathematically, this method computes: <pre>
>   ///   tau = M(q) * vdot + C(q, v) * v
> ```

BTW this comment needs updating to reflect applied forces. Maybe
```
τ = Mv̇+Cv−(f+JᵀF)
```
? Is that the right sign convention? I generally toss the Coriolis & gyroscopic forces in with the other body forces. The only forces I like on the left are constraint forces because it allows for cleaner equations of motion like this:
```
[ M  Gᵀ][v̇]   [f]
[ G  0 ][λ] = [b]
```
where the Coriolis stuff just gets tossed into f and forgotten mostly, and we don't have to sneak in a sign change somewhere for λ.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6958#-:-KtI-xgLA12e8UH2pnuk:b16c6vr)*
<!-- Sent from Reviewable.io -->

siyuanfeng-tri commented:
Reviewed 3 of 7 files at r5.
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[drake/multibody/multibody_tree/mobilizer.h, line 453 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/6958#-KtHmS_RN1DzpTLfLaZ6:-KtILw0EssXrX_u1LyWV:btnaas7) ([raw file](https://github.com/robotlocomotion/drake/blob/78544728340f8de2c30e5c52edd73f9ec81b905c/drake/multibody/multibody_tree/mobilizer.h#L453)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

BTW it was making these methods names way too long IMO. What do you think? especially for the mutable versions like `get_mutable_generalized_forces_from_array()`...
</blockquote></details>

i like generalized better. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6958)*
<!-- Sent from Reviewable.io -->

siyuanfeng-tri commented:
Reviewed 4 of 7 files at r5.
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6958)*
<!-- Sent from Reviewable.io -->

siyuanfeng-tri commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6958#-:-KtISTeakY4RMv24i3RO:bnfp4nl)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
Review status: 8 of 13 files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/common/eigen_types.h, line 238 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/6958#-KtI-itWpePA-wKj26zK:-KtIRiqaIEbPrCT7rDUF:b5zbt5u) ([raw file](https://github.com/robotlocomotion/drake/blob/75c47ffe55f109bd1d7d68a9e0878843c14dd37a/drake/common/eigen_types.h#L238)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW typo
</blockquote></details>

BTW. That was an accident. I removed the change.

---

*[drake/multibody/multibody_tree/mobilizer.h, line 453 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/6958#-KtHmS_RN1DzpTLfLaZ6:-KtISQy0o1FwYz2MnJtn:benp04x) ([raw file](https://github.com/robotlocomotion/drake/blob/78544728340f8de2c30e5c52edd73f9ec81b905c/drake/multibody/multibody_tree/mobilizer.h#L453)):*
<details><summary><i>Previously, siyuanfeng-tri wrote…</i></summary><blockquote>

i like generalized better. 
</blockquote></details>

Ok. Reverted back to `get_generalized_forces()`.

---

*[drake/multibody/multibody_tree/multibody_tree.h, line 734 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/6958#-KtHeGnhcuwjnrrwNCGt:-KtIVrrj0ZFdzGYFsT7c:bax2yze) ([raw file](https://github.com/robotlocomotion/drake/blob/78544728340f8de2c30e5c52edd73f9ec81b905c/drake/multibody/multibody_tree/multibody_tree.h#L734)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

Done. See DRAKE_DEMAND in the cc file. It'd be nice to be able to say `DRAKE_DEMAND(tau_array != nullptr)` @ericcousineau-tri 
</blockquote></details>

This one is blocked on you @siyuanfeng-tri 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6958)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
Reviewed 5 of 5 files at r6.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6958)*
<!-- Sent from Reviewable.io -->

soonho-tri commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> PTAL the comments.

---

Reviewed 4 of 16 files at r1, 1 of 5 files at r2, 1 of 7 files at r4, 2 of 7 files at r5, 5 of 5 files at r6.
Review status: all files reviewed at latest revision, 6 unresolved discussions.

---

*[drake/multibody/benchmarks/acrobot/acrobot.cc, line 140 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/6958#-KtMRkAuNCrowOo_cH-H:-KtMRkAuNCrowOo_cH-I:b797ldo) ([raw file](https://github.com/robotlocomotion/drake/blob/21dc7676ae2fda44a897a7fe3a74b15ccea24eac/drake/multibody/benchmarks/acrobot/acrobot.cc#L140)):*
> ```cpp
> Isometry3<T> Acrobot<T>::CalcElbowOutboardFramePoseInWorldFrame(
>     const T& theta1, const T& theta2) const {
>   // Pose of link2's frame L2cm, a the com, in the world frame.
> ```

BTW, `a the com` -> `at the com`?

---

*[drake/multibody/multibody_tree/body_node.h, line 937 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/6958#-KtIblFYhxPS5apFJl3k:-KtIblFYhxPS5apFJl3l:b-3lwse6) ([raw file](https://github.com/robotlocomotion/drake/blob/0ec3eba238258559c341a031564e7d2d0c2928e6/drake/multibody/multibody_tree/body_node.h#L937)):*
> ```objc
>   Eigen::VectorBlock<VectorX<T>> get_mutable_velocities_from_array(
>       VectorX<T>* v) const {
>     DRAKE_ASSERT(v != nullptr);
> ```

FYI, `DRAKE_ASSERT(v)` should work (ditto below).

---

*[drake/multibody/multibody_tree/mobilizer.h, line 426 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/6958#-KtMSDzxE0FEYqwsiVnt:-KtMSDzxE0FEYqwsiVnu:bi8zpy2) ([raw file](https://github.com/robotlocomotion/drake/blob/21dc7676ae2fda44a897a7fe3a74b15ccea24eac/drake/multibody/multibody_tree/mobilizer.h#L426)):*
> ```objc
>       EigenPtr<VectorX<T>> v_array) const {
>     DRAKE_DEMAND(
>         v_array->size() == this->get_parent_tree().get_num_velocities());
> ```

How about checking `v_array` by having `v_array && v_array->size() == ...`?

---

*[drake/multibody/multibody_tree/multibody_tree.h, line 697 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/6958#-KtI_U3atSvX3_ejLxJa:-KtI_U3atSvX3_ejLxJb:brc91mh) ([raw file](https://github.com/robotlocomotion/drake/blob/21dc7676ae2fda44a897a7fe3a74b15ccea24eac/drake/multibody/multibody_tree/multibody_tree.h#L697)):*
> ```objc
>   ///   To access a mobilizer's reaction force on given body B in this array,
>   ///   use the index returned by Body::get_node_index().
>   ///   Notice that in order to reduce the memory footprint this array can be
> ```

BTW, almost the same sentences are repeated at the below `@note` Is this intended redundancy?

---

*[drake/multibody/multibody_tree/test/double_pendulum_test.cc, line 609 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/6958#-KtMUJBqxHFyQ9CHreNd:-KtMUJBqxHFyQ9CHreNe:b6kfgex) ([raw file](https://github.com/robotlocomotion/drake/blob/21dc7676ae2fda44a897a7fe3a74b15ccea24eac/drake/multibody/multibody_tree/test/double_pendulum_test.cc#L609)):*
> ```cpp
>     // Try first using different arrays for input/ouput:
>     // Initialize output to garbage, it should not affect the results.
>     tau.setOnes() * std::numeric_limits<double>::quiet_NaN();
> ```

It seems that `* std::numeric_limits<double>::quiet_NaN();` has no side effect.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6958#-:-KtMV0pNuVnv7BpoMH-Q:b-ljfbnb)*
<!-- Sent from Reviewable.io -->

EricCousineau-TRI commented:
Review status: all files reviewed at latest revision, 6 unresolved discussions.

---

*[drake/multibody/multibody_tree/test/double_pendulum_test.cc, line 609 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/6958#-KtMUJBqxHFyQ9CHreNd:-KtMVXZZ8o6xStU2NPX7:bo9ifjj) ([raw file](https://github.com/robotlocomotion/drake/blob/21dc7676ae2fda44a897a7fe3a74b15ccea24eac/drake/multibody/multibody_tree/test/double_pendulum_test.cc#L609)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

It seems that `* std::numeric_limits<double>::quiet_NaN();` has no side effect.
</blockquote></details>

Perhaps `tau.setConstant(quiet_NaN())` would work in this place?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6958)*
<!-- Sent from Reviewable.io -->

soonho-tri commented:
Review status: all files reviewed at latest revision, 6 unresolved discussions.

---

*[drake/multibody/multibody_tree/test/double_pendulum_test.cc, line 609 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/6958#-KtMUJBqxHFyQ9CHreNd:-KtMXcwMMep9j06e3iRC:bfao903) ([raw file](https://github.com/robotlocomotion/drake/blob/21dc7676ae2fda44a897a7fe3a74b15ccea24eac/drake/multibody/multibody_tree/test/double_pendulum_test.cc#L609)):*
> Perhaps tau.setConstant(quiet_NaN()) would work in this place?

:+1: (fyi, `fill` should work since it's an alias).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6958)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[drake/multibody/multibody_tree/body_node.h, line 937 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/6958#-KtIblFYhxPS5apFJl3k:-KtMpyMOIo6pWnAQb4TE:b-ec019s) ([raw file](https://github.com/robotlocomotion/drake/blob/0ec3eba238258559c341a031564e7d2d0c2928e6/drake/multibody/multibody_tree/body_node.h#L937)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

FYI, `DRAKE_ASSERT(v)` should work (ditto below).
</blockquote></details>

Excellent!

---

*[drake/multibody/multibody_tree/mobilizer.h, line 426 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/6958#-KtMSDzxE0FEYqwsiVnt:-KtMqD3cW8nPISqJM8pT:b-6fz0u2) ([raw file](https://github.com/robotlocomotion/drake/blob/21dc7676ae2fda44a897a7fe3a74b15ccea24eac/drake/multibody/multibody_tree/mobilizer.h#L426)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

How about checking `v_array` by having `v_array && v_array->size() == ...`?
</blockquote></details>

I added an extra `DRAKE_DEMAND`. I personally like having them separately so that when we get an abort we get something more specific printed to stderr.

---

*[drake/multibody/multibody_tree/multibody_tree.h, line 697 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/6958#-KtI_U3atSvX3_ejLxJa:-KtMqqiZRrs2K0S02KhX:by2dnc7) ([raw file](https://github.com/robotlocomotion/drake/blob/21dc7676ae2fda44a897a7fe3a74b15ccea24eac/drake/multibody/multibody_tree/multibody_tree.h#L697)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

BTW, almost the same sentences are repeated at the below `@note` Is this intended redundancy?
</blockquote></details>

BTW it is not. The `@note` at the end was meant to eliminate redundancy but then clearly I forgot to eliminate this text. Thanks.

---

*[drake/multibody/multibody_tree/test/double_pendulum_test.cc, line 609 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/6958#-KtMUJBqxHFyQ9CHreNd:-KtMrb9bquQUUsaCF8jP:b-64h7o3) ([raw file](https://github.com/robotlocomotion/drake/blob/21dc7676ae2fda44a897a7fe3a74b15ccea24eac/drake/multibody/multibody_tree/test/double_pendulum_test.cc#L609)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

> Perhaps tau.setConstant(quiet_NaN()) would work in this place?

:+1: (fyi, `fill` should work since it's an alias).

</blockquote></details>

doh, good catch! 
BTW, how would you use `fill` here? 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6958)*
<!-- Sent from Reviewable.io -->

soonho-tri commented:
Review status: 6 of 13 files reviewed at latest revision, 3 unresolved discussions.

---

*[drake/multibody/multibody_tree/multibody_tree.h, line 734 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/6958#-KtHeGnhcuwjnrrwNCGt:-KtMtoX8KGD3BTHtY_9q:b-to22fy) ([raw file](https://github.com/robotlocomotion/drake/blob/78544728340f8de2c30e5c52edd73f9ec81b905c/drake/multibody/multibody_tree/multibody_tree.h#L734)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

This one is blocked on you @siyuanfeng-tri 
</blockquote></details>

ping @siyuanfeng-tri...

---

*[drake/multibody/multibody_tree/test/double_pendulum_test.cc, line 609 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/6958#-KtMUJBqxHFyQ9CHreNd:-KtMtrZxha-74GwrYw7-:b-7sqt7p) ([raw file](https://github.com/robotlocomotion/drake/blob/21dc7676ae2fda44a897a7fe3a74b15ccea24eac/drake/multibody/multibody_tree/test/double_pendulum_test.cc#L609)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

doh, good catch! 
BTW, how would you use `fill` here? 
</blockquote></details>

AFAIK, it's just an alias: http://eigen.tuxfamily.org/dox/classEigen_1_1DenseBase.html#a9be169c308801411aa24be93d30930bf

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6958)*
<!-- Sent from Reviewable.io -->

soonho-tri commented:
Reviewed 7 of 7 files at r7.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[drake/multibody/multibody_tree/mobilizer.h, line 426 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/6958#-KtMSDzxE0FEYqwsiVnt:-KtMu74FDKOeXE1APVFg:bs1kjcc) ([raw file](https://github.com/robotlocomotion/drake/blob/21dc7676ae2fda44a897a7fe3a74b15ccea24eac/drake/multibody/multibody_tree/mobilizer.h#L426)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

I added an extra `DRAKE_DEMAND`. I personally like having them separately so that when we get an abort we get something more specific printed to stderr.
</blockquote></details>

LGTM (marked myself satisfied)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6958)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/multibody/multibody_tree/test/double_pendulum_test.cc, line 609 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/6958#-KtMUJBqxHFyQ9CHreNd:-KtMuELcG5nsDho--66k:b-f4m9a3) ([raw file](https://github.com/robotlocomotion/drake/blob/21dc7676ae2fda44a897a7fe3a74b15ccea24eac/drake/multibody/multibody_tree/test/double_pendulum_test.cc#L609)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

AFAIK, it's just an alias: http://eigen.tuxfamily.org/dox/classEigen_1_1DenseBase.html#a9be169c308801411aa24be93d30930bf
</blockquote></details>

oh! I didn't realize that existed at all, I thought you were talking about `std::fill`. Thanks.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6958)*
<!-- Sent from Reviewable.io -->

siyuanfeng-tri commented:
Reviewed 7 of 7 files at r7.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6958)*
<!-- Sent from Reviewable.io -->
